release: prepare for 0.3.0

jonasnick opened the pull request:
Based on #1221. The release date is set to tomorrow March 8th. This is a draft because we only want to merge this tomorrow and  make sure to tag the release immediately afterwards. 

`make dist` does not miss files as shown by this diff between the dist tar.gz and a clean checkout of the repo:
```
❯ diff -qr . ../../../secp256k1-clean
Only in .: aclocal.m4
Only in ./build-aux: ar-lib
Only in ./build-aux: compile
Only in ./build-aux: config.guess
Only in ./build-aux: config.sub
Only in ./build-aux: depcomp
Only in ./build-aux: install-sh
Only in ./build-aux: ltmain.sh
Only in ./build-aux/m4: libtool.m4
Only in ./build-aux/m4: lt~obsolete.m4
Only in ./build-aux/m4: ltoptions.m4
Only in ./build-aux/m4: ltsugar.m4
Only in ./build-aux/m4: ltversion.m4
Only in ./build-aux: missing
Only in ./build-aux: test-driver
Files ./CHANGELOG.md and ../../../secp256k1-clean/CHANGELOG.md differ
Only in ../../../secp256k1-clean: ci
Only in ../../../secp256k1-clean: .cirrus.yml
Only in .: configure
Files ./configure.ac and ../../../secp256k1-clean/configure.ac differ
Only in ../../../secp256k1-clean: .git
Only in ../../../secp256k1-clean: .gitattributes
Only in ../../../secp256k1-clean: .gitignore
Only in .: Makefile.in
```

real-or-random commented:
See https://github.com/bitcoin-core/secp256k1/issues/1176 by the way, but the only thing that's relevant for this release is the check of `make dist`. And that seems good, yes.

real-or-random commented:
It's not about tests and examples but about the changes `secp256k1_context_static` (which have been noted in the Changelog already earlier).

@sipa The reasoning is 
 - The changes to `secp256k1_context_static` are API-incompatible, so normally we would increase MAJOR.
 - But that would make it a 1.0.0. So we make use of https://semver.org/#spec-item-4: "Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable." So if the only goal is to adhere to SemVer 2, we could in fact do whatever we want. But increasing MINOR seems most reasonable for an incompatible change before 1.0.0.

real-or-random commented:
Hm, do you think we should add a bold/whatever note to both of these changes that says that they're API-incompatible?

hebasto commented:
Changes in `configure.ac` need to be mirrored into `CMakeLists.txt`.

jonasnick commented:
Rebased.

> Changes in configure.ac need to be mirrored into CMakeLists.txt.

This will need to be mentioned in the release_process.md. I updated `VERSION` and `LIB_VERSION_*`. Does CMake have an `IS_RELEASE` flag like configure.ac that I need to update as well?

> You're referring to the fact that cloning/destroying secp256k1_context_static is no longer allowed? I guess technically that is an API-incompatible change, but given that it's behavior that we consider almost certainly misuse...

The semantic versioning spec says "Major version X (X.y.z | X > 0) MUST be incremented if **any** backwards incompatible changes are introduced to the public API" (emphasis mine). We're on major version 0 so anything can happen any time, but I'd prefer to have a consistent definition of backward incompatibility. Maybe it's possible to view the changed behavior of cloning/destroying `secp256k1_context_static` as a bug fix, but right now in the CHANGELOG it's clearly documented as a "change".

hebasto commented:
> Does CMake have an `IS_RELEASE` flag like configure.ac that I need to update as well?

No, it has not.

hebasto commented:
From https://github.com/bitcoin-core/secp256k1/pull/1113#issuecomment-1453785807 and https://github.com/bitcoin-core/secp256k1/pull/1113#issuecomment-1453785807:
>    - [ ]  Changelog entry

Sorry for not mentioning it at once.

real-or-random commented:
> From [#1113 (comment)](https://github.com/bitcoin-core/secp256k1/pull/1113#issuecomment-1453785807) and [#1113 (comment)](https://github.com/bitcoin-core/secp256k1/pull/1113#issuecomment-1453785807):
> 
> > * [ ]   Changelog entry
> 
> Sorry for not mentioning it at once.

see #1225

real-or-random commented:
> > Does CMake have an `IS_RELEASE` flag like configure.ac that I need to update as well?
> 
> No, it has not.

Good question! It seems that there's no official flag so far. But fwiw, I've just subscribed to https://gitlab.kitware.com/cmake/cmake/-/issues/16716 and https://gitlab.kitware.com/cmake/cmake/-/issues/23649 which track support for suffixes like `-dev`.

jonasnick commented:
Updated diff between `make dist` and clean checkout:
```
❯ diff -qr . ../../../secp256k1-clean
Only in .: aclocal.m4
Only in ./build-aux: ar-lib
Only in ./build-aux: compile
Only in ./build-aux: config.guess
Only in ./build-aux: config.sub
Only in ./build-aux: depcomp
Only in ./build-aux: install-sh
Only in ./build-aux: ltmain.sh
Only in ./build-aux/m4: libtool.m4
Only in ./build-aux/m4: lt~obsolete.m4
Only in ./build-aux/m4: ltoptions.m4
Only in ./build-aux/m4: ltsugar.m4
Only in ./build-aux/m4: ltversion.m4
Only in ./build-aux: missing
Only in ./build-aux: test-driver
Files ./CHANGELOG.md and ../../../secp256k1-clean/CHANGELOG.md differ
Only in ../../../secp256k1-clean: ci
Only in ../../../secp256k1-clean: .cirrus.yml
Only in ../../../secp256k1-clean: cmake
Only in ../../../secp256k1-clean: CMakeLists.txt
Only in .: configure
Files ./configure.ac and ../../../secp256k1-clean/configure.ac differ
Only in ../../../secp256k1-clean/examples: CMakeLists.txt
Only in ../../../secp256k1-clean: .git
Only in ../../../secp256k1-clean: .gitattributes
Only in ../../../secp256k1-clean: .gitignore
Only in .: Makefile.in
Only in ../../../secp256k1-clean/src: CMakeLists.txt
```

real-or-random reviewed (approved):
ACK 47c7219adaf403d606461bd324e78bdab53ae336

(assuming https://github.com/bitcoin-core/secp256k1/pull/1229 is in)

jonasnick commented:
rebased

real-or-random reviewed (approved):
ACK b40adf23604d4461f6a5a4af984f4d0c2f337933

hebasto reviewed (approved):
ACK b40adf23604d4461f6a5a4af984f4d0c2f337933

6309304695 reviewed (commented):
secp256k1_recovery.h
